README.txt
SpamProbe v1.4 README

LEGALESE

Burton Computer Corporation
http://www.burton-computer.com

Copyright (C) 2002,2003,2004,2005,2006 Burton Computer Corporation ALL
RIGHTS RESERVED

This program is open source software; you can redistribute it and/or
modify it under the terms of the Q Public License (QPL) version
1.0. Use of this software in whole or in part, including linking it
(modified or unmodified) into other programs is subject to the terms
of the QPL.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the Q Public
License for more details.

You should have received a copy of the Q Public License along with
this program; see the file LICENSE.txt.  If not, visit the Burton
Computer Corporation or CoolDevTools web site QPL pages at:

    http://www.burton-computer.com/qpl.html


INTRODUCTION

Welcome to SpamProbe!  Are you tired of the constant bombardment of
your inbox by unwanted email pushing everything from porn to get rich
quick schemes?  Have you tried other spam filters but become
disenchanted with them when you realized that their manually generated
rule sets weren't updated fast enough to keep up with spammers wording
changes?  Or that they generated unwanted false positive scores?

SpamProbe operates on a different basis entirely.  Instead of using
pattern matching and a set of human generated rules SpamProbe relies
on a Bayesian analysis of the frequency of words used in spam and
non-spam emails received by an individual person.  The process is
completely automatic and tailors itself to the kinds of emails that
each person receives.

My work on SpamProbe was inspired by an excellent article by Paul
Graham.  He describes the basic idea and his results.  You can read
his article here:

  http://www.paulgraham.com/spam.html

I highly recommend reading the article and the other spam related
links on his site for excellent insights into why spam is a problem
and how you can defeat it.  Of course running SpamProbe is an
excellent step! :-)


FEATURES

 * Spam detection using Bayesian analysis of terms contained in each
   email.  Words used often in spams but not in good email tend to
   indicate that a message is spam.

 * Written in C++ for good performance.  Database access using PBL or
   BerkeleyDB for quick startup and fast term count retrieval.

 * Recognition and decoding of MIME attachments in quoted-printable
   and base64 encoding.  Automatically skips non-text attachments.
   MIME decoding enables SpamProbe to make decisions based on words in
   the emails rather than base64 gobbledigook.

 * Counts two word phrases as well as single words for higher
   precision.

 * Counts words in from, to, cc, received, and subject headers
   separately for improved accuracy since some words are better
   indicators of spam when they appear in the subject than in the
   body.  Received headers after the first are counted separately from
   first Received header since they are likely to be forged in spams.

 * Ignores HTML tags in emails for scoring purposes unless the -h
   command line option is used.  Many spams use HTML and few humans do
   so HTML tends to become a powerful recognizer of spams.  However in
   the author's opinion this also substantially increases the
   likelihood of false positives if someone does send a non-spam email
   containing HTML tags.  SpamProbe does pull urls from inside of html
   tags however since those tend to be spammer specific.

 * Supports MBOX, MBX and Maildir mailbox formats.

 * Analyzes image attachments to extract useful terms for scoring.
   This catches the majority of spams with only images and no words.

 * Locks mboxes and databases using fcntl file locking to avoid
   problems when multiple emails arrive simultaneously.  Can
   optionally be compiled to use Berkeley DB's environment functions
   to provide better interaction with db_dump, db_verify, etc.

 * Scores only the Received, Subject, To, From, and Cc headers.  All
   other headers are ignored to make it hard for spammers to hide
   non-spammy words in X- headers to fool the filter.  The -H command
   line option can be used to override this.

 * Supports Content-Length: field in mbox headers.  This can be
   disabled using -Y option to use only From_ to recognize new
   messages.

 * Uses MD5 digest of emails to recognize reclassification of an
   already classified spam to avoid distortion of the word counts if
   emails are reclassified.  This way emails can be kept in an mbox
   that is repeatedly scanned by spamprobe without counting them more
   than once.  Digest printed along with score can be stored in emails
   so that binary attachments can be removed by MUA after scoring
   without breaking SpamProbe's ability to reclassify the email later.

 * Provides a time stamp based database cleanup command to remove
   terms from the database if their counts never rise above a certain
   threshold value (normally 2).  Tends to limit the otherwise
   unbounded growth of the database as each message adds some unique
   terms.  Also provides a purge command to remove all terms with
   counts below a specified minimum no matter what their age.

 * The -X option uses a more aggressive scoring algorithm which uses
   all significantly good or spammy terms when scoring a message and
   also gives more weight to within message frequency of terms.  This
   method improves recall and works best when the database contains a
   few thousand messages.

 * edit-term command allows users to directly modify the counts of
   individual terms.  For example to force a particular term to be
   considered spammy or good.

 * train mode can be used in place of receive to minimize database
   writes by only updating database for difficult to classify
   messages.

 * auto-train mode can be used to build a database efficiently from a
   collection of email stored in multiple files.

 * Multiple database options including BerkeleyDB, PBL, and fixed size
   hash file.


INSTALLATION

SpamProbe is distributed in source code form.  The current version
uses autoconf to handle generating an appropriate Makefile for your
system.  Generally you should be able to install SpamProbe like this:

  ./configure
  make
  make install

The configure script accepts some useful command line options
including:

 --prefix=/path/to/install

   Use this to specify a directory tree to install SpamProbe and its
   man page into.  The default is /usr/local.

 --with-pbl=/path/to/pbl

   Use this to specify the directory where you installed PBL (Peter
   Graf's Program Base Library) if the configure script is not able to
   find it on its own.  See notes below for more information about the
   relative merits of PBL and Berkeley DB.

 --with-db=/path/to/db

   If you would rather use Berkeley DB than PBL use this to specify
   the directory where you installed Berkeley DB if the configure
   script is not able to find it on its own.

 --enable-cdb

   Use this to enable use of Berkeley DB's CDB mode.  In this mode SP
   will create and use a Berkeley DB "environment" and take advantage
   of the "concurrent data store" feature of Berkeley DB.
   Theoretically this allows better interaction with Berkeley DB's
   utilities (db_verify, db_recover, etc) but some people have
   reported problems with this mode so it is now optional.

 --enable-default-8bit

   Use this if you want to change SP's default handling of eight bit
   characters.  Normally 8 bit characters are converted to the letter
   z in the database to save space (this can be changed at runtime
   using the -8 command line option).  If you normally receive 8 bit
   characters in email then you might want to compile with this flag
   to avoid the need to remember to type -8 all of the time.

 --enable-assert

   Use this if you want to enable the assertions in the source code
   for debugging purposes.  Generally this is not desirable for
   production use since the assertions reduce performance.

 --enable-big-endian

   Causes all database reads and writes to use big-endian byte order
   even on little-endian platforms.  Use this if you need to copy a
   database across different platforms and want to guarantee byte
   level data portability.  This option slows down database I/O on
   little endian systems.

If you want to statically link the BerkeleyDB library run configure
with the appropriate command line option in the LDFLAGS variable.  For
gcc you do could something like this:

  LDFLAGS=-static ./configure
  make
  make install

If you want to use Berkeley DB and configure does not find your
BerkeleyDB library and you know it is installed but has a non-standard
name then you can explicitly define the library to use.  For example
some RedHat 6.2 systems might require:

  CPPFLAGS=-I/usr/include/db3 ./configure

You need GNU G++ or some other C++ compiler to compile SpamProbe.  You
also need either PBL or Berkeley DB installed on your system.  PBL
uses more disk space and may require you to download and install it on
your system (see below) but is preferred because it supports
transactions and is faster under heavy loads.

Berkeley DB comes preinstalled on some operating systems so using it
may be easier.  It also creates smaller data files than PBL.  However
some users have reported data corruption issues with Berkeley DB in
the past so use of PBL is recommended.  That having been said the
author of SpamProbe used Berkeley DB for almost a year without any
problems in production so the reported issues seem to be OS or system
specific.


BUILD NOTES USING PBL

To obtain PBL download it from Peter Graf's web site:

  http://mission.base.com/peter/source/

PBL does not use configure for its build process however its makefile
seems to be appropriate out of the box for Red Hat Linux systems and
may work without changes on other posix-like systems as well.  To
compile SpamProbe with PBL I did the following:

  <downloaded pbl_1_03.tar.gz to the current directory>
  tar xzvf pbl_1_03.tar.gz
  cd pbl_1_03
  make
  cd ..
  <downloaded spamprobe-1.0.tar.gz to the current directory>
  tar xzvf spamprobe-1.0.tar.gz
  cd spamprobe-1.0
  ./configure --with-pbl=../pbl_1_03
  make
  make install


BUILD NOTES USING BERKELEY DB

To obtain BerkeleyDB if you do not already have it installed (many
Linux distributions come with DB preinstalled) visit the SleepCat
website to download a copy.  BerkeleyDB is open source and you can use
it for free.

  http://www.sleepycat.com

Installation instructions for BerkeleyDB can be found here:

  http://www.sleepycat.com/docs/ref/build_unix/intro.html


NOTES USING HASH DATABASE

SpamProbe can use a simple, fixed size hash data file as an
alternative to PBL or BDB.  There are two advantages to the hash
format.  The first is speed.  In my experiments the hash file format
is around 2x the speed of PBL (ranged from 1.8x to 3.5x). The second
advantage is that the hash data file size is fixed.  You choose a size
when you create the file and it never changes.  File size can be
anywhere from 1-100 MB. You need to choose a size large enough to hold
your terms with room to spare.  More on that later.

The hash file format also has significant disadvantages.  Becuase the
file size is fixed you must monitor the file to ensure that it does
not become overly full.  When the file becomes more than half full
performance will suffer.  Also the hash format does not store original
terms so you cannot use the dump command to learn what terms are
spammy or hammy in your database.  Finally, the hash format is
imprecise.  Hash collisions can cause the counts from different terms
to be mixed together which can reduce accuracy.

To create a hash data file you add a prefix to the directory name in
the -d command line option.  You can specify just the directory like
this:

  spamprobe -d hash:$HOME/.spamprobe

or you can add a size in megabytes for the file like this:

  spamprobe -d hash:42:$HOME/.spamprobe

The size is only used when a file is first created.  SP auto detects
the size of an existing hash file.  You need to allow enough space for
twice as many terms as you are likely to have in your file.  In my
database I have 2.2 million terms.  That required a database of are 53
MB.  SP uses 12 bytes per term in the hash file so you can estimate
the file size you'll need by multiplying the number of terms by 24.

The hash format does not store the original terms.  Instead it stores
the 32 bit hash code for each term.  You can do just about anything
with a hash file that you could with a PBL file including
import/export, edit-term, cleanup, purge, etc.  You can use export
your PBL database and import it to build a hash file (note that you
cannot go the other direction) and you can export one hash file and
import into a new one to enlarge your file.


SETUP OF SPAMPROBE FOR USERS

Once you have a spamprobe executable copy it to someplace in your PATH
so that procmail can find it.  Then create a directory for SpamProbe
to store its databases in.  By default SpamProbe wants to use the
directory ~/.spamprobe.  You must create this directory manually in
order to run SpamProbe or else specify some other directory using the
-d option.  Something like this should suffice:

  spamprobe create-db

or to use a different directory add the -d command line option:

  spamprobe -d somedirectory create-db

SpamProbe can use either the PBL or Berkeley DB library for its
databases.  Both are fast on local file systems but very slow over
NFS.  Please ensure that your spamprobe directory is on a local file
system to ensure good performance.


MAILDIR FORMAT

SpamProbe will accept a maildir directory name anywhere that an Mbox
or MBX file name can be specified.  When SpamProbe encounters a
Maildir mailbox (directory) name it will automatically process all of
the non-hidden files in the cur and new subdirectories of the mailbox.
There is no need to individually specify these subdirectories.


GETTING STARTED

SpamProbe is not a stand alone mail filter.  It doesn't sort your mail
or split it into different mailboxes.  Instead it relies on some other
program such as procmail to actually file your mail for you.  What
SpamProbe does do is track the word counts in good and spam emails and
generate a score for each email that indicates whether or not it is
likely to be spam.  Scores range from 0 to 1 with any score of 0.9 or
higher indicating a probable spam.

Personally I use SpamProbe with procmail to filter my incoming email
into mail boxes.  I have procmail score each inbound email using
SpamProbe and insert a special header into each email containing its
score.  Then I have procmail move spams into a special mailbox.

No spam filter is perfect and SpamProbe sometimes makes mistakes.  To
correct those mistakes I have a special mailbox that I put undetected
spams into.  I run SpamProbe periodically and have it reclassify any
emails in that mailbox as spam so that it will make a better guess the
next time around.


ChangeLog
2006-11-16  Brian Burton  <brian@burton-computer.com>

	* Released as 1.4d

	* configure.ac: Added ability to selectively disable image
	  processing using --without-gif, --without-jpeg, and/or
	  --without-png.

	* src/spamprobe/spamprobe.cc (set_headers): Added ability to
	  selectively ignore individual headers using -H-headername.

	* src/includes/Ptr,Ref,Array.h: Restored missing <cassert> include.

	* src/parser/PngParser.cc (tokenizeImage): added basic tokens from
	  PNG images.

2006-11-16  Brian Burton  <brian@localhost.localdomain>

	* src/parser/PngParser.cc (PngParser): Stub for PNG parsing using
	  libpng.

	* src/parser/JpegParser.cc (tokenizeMarker): Preliminary
	  implementation of jpeg parsing using jpeglib.

	* configure.ac: Auto detect of either libungig or libgif depending
	  on which one is available.

2007-01-04  Brian Burton  <brian@burton-computer.com>

	* Released as 1.4c

	* spamprobe.1: Modified man page to remove unnecessary informaton
	  and make it more conformant with man page conventions.

	* src/spamprobe/spamprobe.cc (process_extended_options): added
	  ignore-body option.

	* src/parser/HeaderPrefixList.cc (HeaderPrefixList::addHeaderPrefix):
	  Forced header prefixes and names to lower case instead of
	  relying on an assert to enforce the restriction.

	* src/database/FrequencyDBImpl_hash.cc (hash::FrequencyDBImpl_hash): 
	  Disabled experimental hash database auto-cleaning.

	* src/includes/Ref.h: Removed cassert include.

	* src/spamprobe/spamprobe.cc (process_extended_options): Added
	  whitelist option to allow use of SP as a bayesian white list in
	  conjunction with other filters.

2006-02-17  Brian Burton  <brian@burton-computer.com>

	* Released as 1.4b

	* src/parser/TraditionalMailMessageParser.cc (parseBody): Fixed
	  bug reported by Jphn Chandler that prevented tokens from being
	  extracted from headers of messages with no body.

	* src/input/SimpleMultiLineStringCharReader.cc (class
	  SimpleMultiLineStringCharReaderPosition): Fixed crash bug
	  reported by David Rosen when a missing base64 encoded body was
	  parsed.

2006-01-30  Brian Burton  <brian@burton-computer.com>

	* Released as 1.4a

2006-01-28  Brian Burton  <brian@burton-computer.com>

	* src/includes/LRUCache.h (LRUCache): Reimplemented using STL list
	  class.  Map uses Node ptr as key to avoid having two copies of
	  the key in memory.  Changed iterators to use STL style syntax.

	* src/parser/PhrasingTokenizer.cc (compactChars): Improved
	  efficiency.

2006-01-25  Brian Burton  <brian@burton-computer.com>

	* src/parser/MimeDecoder.cc (next_char64): Fixed potential array
	  bounds overflow bug in base64 decoding.  Thanks to Chris Ross
	  for the bug report.

        * src/spamprobe.cc: (and various other files) Added
	  min_phrase_chars option that causes parser to keep adding tokens
	  for phrases until they are at least min_phrase_chars long
	  instead of stopping at phrase word limit.  This might be useful
	  for catching "v i a g r a" as a single term.  So far though
	  experiments with using this option are not very promising.
	
	* src/spamprobe/Command_auto_train.cc (execute): Added LOG
	  sub-command to log each processed message to stdout along with
	  whether or not it had been scored successfully prior to
	  training.  This is useful for experiments to determine how fast
	  SP can learn.

	* src/spamprobe/Command_receive.cc (createScoreCommand): Fixed
	  documentation bug and improved online help for score command.
	  Thanks to Chris Ross for the bug report.

2006-01-03  Brian Burton  <brian@burton-computer.com>

	* src/parser/MailMessage.cc (MailMessage): Removed redundant
	  bounds check.

	* src/utility/MultiLineString.cc (line): Added bounds check.

	* src/parser/HtmlTokenizer.cc (tokenize): Added TempPtr to ensure
	  reader and receiver are cleared between runs.

	* src/utility/MultiLineSubString.cc (m_target): Fixed boundary
	  conditions if passed in indexes are outside bounds of target.

2005-12-28  Brian Burton  <brian@burton-computer.com>

        * Released as 1.4.

	* src/includes/LRUPtrCache.h (LRUPtrCache): Fixed clear() bug that
	  caused -P command line option to crash on some architectures due
	  to an invalid delete.

	* src/database/FrequencyDBImpl_bdb.cc (openDatabase): Fixed
	  relative path database opening bug in CDB mode.  (Thanks to
	  Nicolas Duboc for report and suggested fix).

	* src/spamprobe/AbstractCommand.cc (openDatabase): All non-read
	  only commands will now create the database directory if it's
	  missing when they open the database for the first time and
	  the -c option was specified on the command line.

	* src/database/HashDataFile.cc (close): Replaced clear() with
	  erase().

	* src/spamprobe/Command_help.cc (printCommandLineOptions): Added
	  comprehensive command line option help.

	* src/input/IstreamCharReader.cc (forward): Now uses rdbuf
	  directly when reading characters and seeking to eliminate extra
	  overhead of istream class.

	* src/spamprobe/spamprobe.cc (main): -V option now returns exit
	  code 0 instead of 1.

2005-12-25  Brian Burton  <brian@burton-computer.com>

	* src/database/FrequencyDBImpl_cache.cc (writeWord): Modified
	  database cache to keep modified/unmodified terms within cache
	  size limit.
	  (flush): number of records written to disk per transaction
	  limited to avoid problem with PBL using excessive amounts of
	  memory to write a large cache to disk if the database is large.

2005-12-23  Brian Burton  <brian@burton-computer.com>

	* src/database/DatabaseConfig.cc (createDatabaseImpl): Restored
	  use of database cache.

2005-12-21  Brian Burton  <brian@burton-computer.com>

	* src/includes/Ref.h (T>): Eliminated RefObject base class.

2005-12-18  Brian Burton  <brian@burton-computer.com>

	* src/utility/util.cc (to_7bits): Removed use of string::+=

	* src/input/LineReader.cc (forward): Removed use of string::+=

	* src/utility/MultiLineString.cc (parseText): Fixed bug that

src/utility/md5.c
/*
  Copyright (C) 1999, 2000, 2002 Aladdin Enterprises.  All rights reserved.

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.

  L. Peter Deutsch
  ghost@aladdin.com

 */
/* Id: md5.c,v 1.6 2002/04/13 19:20:28 lpd Exp  */
/*
  Independent implementation of MD5 (RFC 1321).

  This code implements the MD5 Algorithm defined in RFC 1321, whose
  text is available at
	http://www.ietf.org/rfc/rfc1321.txt
  The code is derived from the text of the RFC, including the test suite
  (section A.5) but excluding the rest of Appendix A.  It does not include
  any code or documentation that is identified in the RFC as being
  copyrighted.

  The original and principal author of md5.c is L. Peter Deutsch
  <ghost@aladdin.com>.  Other authors are noted in the change history
  that follows (in reverse chronological order):

  2002-04-13 lpd Clarified derivation from RFC 1321; now handles byte order
	either statically or dynamically; added missing #include <string.h>
	in library.
  2002-03-11 lpd Corrected argument list for main(), and added int return
	type, in test program and T value program.
  2002-02-21 lpd Added missing #include <stdio.h> in test program.
  2000-07-03 lpd Patched to eliminate warnings about "constant is
	unsigned in ANSI C, signed in traditional"; made test program
	self-checking.
  1999-11-04 lpd Edited comments slightly for automatic TOC extraction.
  1999-10-18 lpd Fixed typo in header comment (ansi2knr rather than md5).
  1999-05-03 lpd Original version.
 */

#include "md5.h"
#include <string.h>

#undef BYTE_ORDER	/* 1 = big-endian, -1 = little-endian, 0 = unknown */
#ifdef ARCH_IS_BIG_ENDIAN
#  define BYTE_ORDER (ARCH_IS_BIG_ENDIAN ? 1 : -1)
#else
#  define BYTE_ORDER 0
#endif

#define T_MASK ((md5_word_t)~0)
#define T1 /* 0xd76aa478 */ (T_MASK ^ 0x28955b87)
#define T2 /* 0xe8c7b756 */ (T_MASK ^ 0x173848a9)
#define T3    0x242070db
#define T4 /* 0xc1bdceee */ (T_MASK ^ 0x3e423111)
#define T5 /* 0xf57c0faf */ (T_MASK ^ 0x0a83f050)
#define T6    0x4787c62a
#define T7 /* 0xa8304613 */ (T_MASK ^ 0x57cfb9ec)
#define T8 /* 0xfd469501 */ (T_MASK ^ 0x02b96afe)
#define T9    0x698098d8
#define T10 /* 0x8b44f7af */ (T_MASK ^ 0x74bb0850)
#define T11 /* 0xffff5bb1 */ (T_MASK ^ 0x0000a44e)
#define T12 /* 0x895cd7be */ (T_MASK ^ 0x76a32841)
#define T13    0x6b901122
#define T14 /* 0xfd987193 */ (T_MASK ^ 0x02678e6c)
#define T15 /* 0xa679438e */ (T_MASK ^ 0x5986bc71)
#define T16    0x49b40821
#define T17 /* 0xf61e2562 */ (T_MASK ^ 0x09e1da9d)
#define T18 /* 0xc040b340 */ (T_MASK ^ 0x3fbf4cbf)
#define T19    0x265e5a51
#define T20 /* 0xe9b6c7aa */ (T_MASK ^ 0x16493855)
#define T21 /* 0xd62f105d */ (T_MASK ^ 0x29d0efa2)
#define T22    0x02441453
#define T23 /* 0xd8a1e681 */ (T_MASK ^ 0x275e197e)
#define T24 /* 0xe7d3fbc8 */ (T_MASK ^ 0x182c0437)
#define T25    0x21e1cde6
#define T26 /* 0xc33707d6 */ (T_MASK ^ 0x3cc8f829)
#define T27 /* 0xf4d50d87 */ (T_MASK ^ 0x0b2af278)
#define T28    0x455a14ed
#define T29 /* 0xa9e3e905 */ (T_MASK ^ 0x561c16fa)
#define T30 /* 0xfcefa3f8 */ (T_MASK ^ 0x03105c07)
#define T31    0x676f02d9
#define T32 /* 0x8d2a4c8a */ (T_MASK ^ 0x72d5b375)
#define T33 /* 0xfffa3942 */ (T_MASK ^ 0x0005c6bd)
#define T34 /* 0x8771f681 */ (T_MASK ^ 0x788e097e)
#define T35    0x6d9d6122
#define T36 /* 0xfde5380c */ (T_MASK ^ 0x021ac7f3)
#define T37 /* 0xa4beea44 */ (T_MASK ^ 0x5b4115bb)
#define T38    0x4bdecfa9
#define T39 /* 0xf6bb4b60 */ (T_MASK ^ 0x0944b49f)
#define T40 /* 0xbebfbc70 */ (T_MASK ^ 0x4140438f)
#define T41    0x289b7ec6
#define T42 /* 0xeaa127fa */ (T_MASK ^ 0x155ed805)
#define T43 /* 0xd4ef3085 */ (T_MASK ^ 0x2b10cf7a)
#define T44    0x04881d05
#define T45 /* 0xd9d4d039 */ (T_MASK ^ 0x262b2fc6)
#define T46 /* 0xe6db99e5 */ (T_MASK ^ 0x1924661a)
#define T47    0x1fa27cf8
#define T48 /* 0xc4ac5665 */ (T_MASK ^ 0x3b53a99a)
#define T49 /* 0xf4292244 */ (T_MASK ^ 0x0bd6ddbb)
#define T50    0x432aff97
#define T51 /* 0xab9423a7 */ (T_MASK ^ 0x546bdc58)
#define T52 /* 0xfc93a039 */ (T_MASK ^ 0x036c5fc6)
#define T53    0x655b59c3
#define T54 /* 0x8f0ccc92 */ (T_MASK ^ 0x70f3336d)
#define T55 /* 0xffeff47d */ (T_MASK ^ 0x00100b82)
#define T56 /* 0x85845dd1 */ (T_MASK ^ 0x7a7ba22e)
#define T57    0x6fa87e4f
#define T58 /* 0xfe2ce6e0 */ (T_MASK ^ 0x01d3191f)
#define T59 /* 0xa3014314 */ (T_MASK ^ 0x5cfebceb)
#define T60    0x4e0811a1
#define T61 /* 0xf7537e82 */ (T_MASK ^ 0x08ac817d)
#define T62 /* 0xbd3af235 */ (T_MASK ^ 0x42c50dca)
#define T63    0x2ad7d2bb
#define T64 /* 0xeb86d391 */ (T_MASK ^ 0x14792c6e)


static void
md5_process(md5_state_t *pms, const md5_byte_t *data /*[64]*/)
{
    md5_word_t
	a = pms->abcd[0], b = pms->abcd[1],
	c = pms->abcd[2], d = pms->abcd[3];
    md5_word_t t;
#if BYTE_ORDER > 0
    /* Define storage only for big-endian CPUs. */
    md5_word_t X[16];
#else
    /* Define storage for little-endian or both types of CPUs. */
    md5_word_t xbuf[16];
    const md5_word_t *X;
#endif

    {
#if BYTE_ORDER == 0
	/*
	 * Determine dynamically whether this is a big-endian or
	 * little-endian machine, since we can use a more efficient
	 * algorithm on the latter.
	 */
	static const int w = 1;

	if (*((const md5_byte_t *)&w)) /* dynamic little-endian */
#endif
#if BYTE_ORDER <= 0		/* little-endian */
	{
	    /*
	     * On little-endian machines, we can process properly aligned
	     * data without copying it.
	     */
	    if (!((data - (const md5_byte_t *)0) & 3)) {
		/* data are properly aligned */
		X = (const md5_word_t *)data;
	    } else {
		/* not aligned */
		memcpy(xbuf, data, 64);
		X = xbuf;
	    }
	}
#endif
#if BYTE_ORDER == 0
	else			/* dynamic big-endian */
#endif
#if BYTE_ORDER >= 0		/* big-endian */
	{
	    /*
	     * On big-endian machines, we must arrange the bytes in the
	     * right order.
	     */
	    const md5_byte_t *xp = data;
	    int i;

#  if BYTE_ORDER == 0
	    X = xbuf;		/* (dynamic only) */
#  else
#    define xbuf X		/* (static only) */
#  endif
	    for (i = 0; i < 16; ++i, xp += 4)
		xbuf[i] = xp[0] + (xp[1] << 8) + (xp[2] << 16) + (xp[3] << 24);
	}
#endif
    }

#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32 - (n))))

    /* Round 1. */
    /* Let [abcd k s i] denote the operation
       a = b + ((a + F(b,c,d) + X[k] + T[i]) <<< s). */
#define F(x, y, z) (((x) & (y)) | (~(x) & (z)))
#define SET(a, b, c, d, k, s, Ti)\
  t = a + F(b,c,d) + X[k] + Ti;\
  a = ROTATE_LEFT(t, s) + b
    /* Do the following 16 operations. */
    SET(a, b, c, d,  0,  7,  T1);
    SET(d, a, b, c,  1, 12,  T2);
    SET(c, d, a, b,  2, 17,  T3);
    SET(b, c, d, a,  3, 22,  T4);
    SET(a, b, c, d,  4,  7,  T5);
    SET(d, a, b, c,  5, 12,  T6);
    SET(c, d, a, b,  6, 17,  T7);
    SET(b, c, d, a,  7, 22,  T8);
    SET(a, b, c, d,  8,  7,  T9);
    SET(d, a, b, c,  9, 12, T10);
    SET(c, d, a, b, 10, 17, T11);
    SET(b, c, d, a, 11, 22, T12);
    SET(a, b, c, d, 12,  7, T13);
    SET(d, a, b, c, 13, 12, T14);
    SET(c, d, a, b, 14, 17, T15);
    SET(b, c, d, a, 15, 22, T16);
#undef SET

     /* Round 2. */
     /* Let [abcd k s i] denote the operation
          a = b + ((a + G(b,c,d) + X[k] + T[i]) <<< s). */
#define G(x, y, z) (((x) & (z)) | ((y) & ~(z)))
#define SET(a, b, c, d, k, s, Ti)\
  t = a + G(b,c,d) + X[k] + Ti;\
  a = ROTATE_LEFT(t, s) + b
     /* Do the following 16 operations. */
    SET(a, b, c, d,  1,  5, T17);
    SET(d, a, b, c,  6,  9, T18);
    SET(c, d, a, b, 11, 14, T19);
    SET(b, c, d, a,  0, 20, T20);
    SET(a, b, c, d,  5,  5, T21);
    SET(d, a, b, c, 10,  9, T22);
    SET(c, d, a, b, 15, 14, T23);
    SET(b, c, d, a,  4, 20, T24);
    SET(a, b, c, d,  9,  5, T25);
    SET(d, a, b, c, 14,  9, T26);
    SET(c, d, a, b,  3, 14, T27);
    SET(b, c, d, a,  8, 20, T28);
    SET(a, b, c, d, 13,  5, T29);
    SET(d, a, b, c,  2,  9, T30);
    SET(c, d, a, b,  7, 14, T31);
    SET(b, c, d, a, 12, 20, T32);
#undef SET

     /* Round 3. */
     /* Let [abcd k s t] denote the operation
          a = b + ((a + H(b,c,d) + X[k] + T[i]) <<< s). */
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define SET(a, b, c, d, k, s, Ti)\
  t = a + H(b,c,d) + X[k] + Ti;\
  a = ROTATE_LEFT(t, s) + b
     /* Do the following 16 operations. */
    SET(a, b, c, d,  5,  4, T33);
    SET(d, a, b, c,  8, 11, T34);
    SET(c, d, a, b, 11, 16, T35);
    SET(b, c, d, a, 14, 23, T36);
    SET(a, b, c, d,  1,  4, T37);
    SET(d, a, b, c,  4, 11, T38);
    SET(c, d, a, b,  7, 16, T39);
    SET(b, c, d, a, 10, 23, T40);
    SET(a, b, c, d, 13,  4, T41);
    SET(d, a, b, c,  0, 11, T42);
    SET(c, d, a, b,  3, 16, T43);
    SET(b, c, d, a,  6, 23, T44);
    SET(a, b, c, d,  9,  4, T45);
    SET(d, a, b, c, 12, 11, T46);
    SET(c, d, a, b, 15, 16, T47);
    SET(b, c, d, a,  2, 23, T48);
#undef SET

     /* Round 4. */
     /* Let [abcd k s t] denote the operation
          a = b + ((a + I(b,c,d) + X[k] + T[i]) <<< s). */
#define I(x, y, z) ((y) ^ ((x) | ~(z)))
#define SET(a, b, c, d, k, s, Ti)\
  t = a + I(b,c,d) + X[k] + Ti;\
  a = ROTATE_LEFT(t, s) + b
     /* Do the following 16 operations. */
    SET(a, b, c, d,  0,  6, T49);
    SET(d, a, b, c,  7, 10, T50);
    SET(c, d, a, b, 14, 15, T51);
    SET(b, c, d, a,  5, 21, T52);
    SET(a, b, c, d, 12,  6, T53);
    SET(d, a, b, c,  3, 10, T54);
    SET(c, d, a, b, 10, 15, T55);
    SET(b, c, d, a,  1, 21, T56);
    SET(a, b, c, d,  8,  6, T57);
    SET(d, a, b, c, 15, 10, T58);
    SET(c, d, a, b,  6, 15, T59);
    SET(b, c, d, a, 13, 21, T60);
    SET(a, b, c, d,  4,  6, T61);
    SET(d, a, b, c, 11, 10, T62);
    SET(c, d, a, b,  2, 15, T63);
    SET(b, c, d, a,  9, 21, T64);
#undef SET

     /* Then perform the following additions. (That is increment each
        of the four registers by the value it had before this block
        was started.) */
    pms->abcd[0] += a;
    pms->abcd[1] += b;
    pms->abcd[2] += c;
    pms->abcd[3] += d;
}

void
md5_init(md5_state_t *pms)
{
    pms->count[0] = pms->count[1] = 0;
    pms->abcd[0] = 0x67452301;
    pms->abcd[1] = /*0xefcdab89*/ T_MASK ^ 0x10325476;
    pms->abcd[2] = /*0x98badcfe*/ T_MASK ^ 0x67452301;
    pms->abcd[3] = 0x10325476;
}

void
md5_append(md5_state_t *pms, const md5_byte_t *data, int nbytes)
{
    const md5_byte_t *p = data;
    int left = nbytes;
    int offset = (pms->count[0] >> 3) & 63;
    md5_word_t nbits = (md5_word_t)(nbytes << 3);

    if (nbytes <= 0)
	return;

    /* Update the message length. */
    pms->count[1] += nbytes >> 29;
    pms->count[0] += nbits;
    if (pms->count[0] < nbits)
	pms->count[1]++;

    /* Process an initial partial block. */
    if (offset) {
	int copy = (offset + nbytes > 64 ? 64 - offset : nbytes);

	memcpy(pms->buf + offset, p, copy);
	if (offset + copy < 64)
	    return;
	p += copy;
	left -= copy;
	md5_process(pms, pms->buf);
    }

    /* Process full blocks. */
    for (; left >= 64; p += 64, left -= 64)
	md5_process(pms, p);

    /* Process a final partial block. */
    if (left)
	memcpy(pms->buf, p, left);
}

void
md5_finish(md5_state_t *pms, md5_byte_t digest[16])
{
    static const md5_byte_t pad[64] = {
	0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
    };
    md5_byte_t data[8];
    int i;

    /* Save the length before padding. */
    for (i = 0; i < 8; ++i)
	data[i] = (md5_byte_t)(pms->count[i >> 2] >> ((i & 3) << 3));
    /* Pad to 56 bytes mod 64. */
    md5_append(pms, pad, ((55 - (pms->count[0] >> 3)) & 63) + 1);
    /* Append the length. */
    md5_append(pms, data, 8);
    for (i = 0; i < 16; ++i)
	digest[i] = (md5_byte_t)(pms->abcd[i >> 2] >> ((i & 3) << 3));
}
